skyrim-platform/src/platform_se/skyrim_platform/GameEventSinks.cpp

```diff
@@ -665,3 +732,1297 @@ RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
+        return;
+      }
+
+      obj.SetProperty("target", CreateObject("ObjectReference", targetLocal));
+      obj.SetProperty("cause", CreateObject("ObjectReference", causeLocal));
+
+      EventsApi::SendEvent("triggerLeave", { JsValue::Undefined(), obj });
+    });
+
+  return RE::BSEventNotifyControl::kContinue;
+}
+
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
+  const RE::TESSleepStartEvent* event,
+  RE::BSTEventSource<RE::TESSleepStartEvent>* eventSource)
+{
+  auto converted =
+    reinterpret_cast<const TESEvents::TESSleepStartEvent*>(event);
+
+  if (!converted) {
+    return RE::BSEventNotifyControl::kContinue;
+  }
+
+  auto sleepStart = converted->sleepStartTime;
+  auto sleepEnd = converted->desiredSleepEndTime;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([sleepStart, sleepEnd] {
+    auto obj = JsValue::Object();
+
+    if (sleepStart < 0 || sleepEnd < 0) {
+      return;
+    }
+
+    obj.SetProperty("startTime", JsValue::Double(sleepStart));
+    obj.SetProperty("desiredEndTime", JsValue::Double(sleepEnd));
+
+    EventsApi::SendEvent("sleepStart", { JsValue::Undefined(), obj });
+  });
+
+  return RE::BSEventNotifyControl::kContinue;
+}
+
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
+  const RE::TESSleepStopEvent* event,
+  RE::BSTEventSource<RE::TESSleepStopEvent>* eventSource)
+{
+  auto converted =
+    reinterpret_cast<const TESEvents::TESSleepStopEvent*>(event);
+
+  if (!converted) {
+    return RE::BSEventNotifyControl::kContinue;
+  }
+
+  auto isInterrupted = converted->isInterrupted;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask([isInterrupted] {
+    auto obj = JsValue::Object();
+
+    obj.SetProperty("isInterrupted", JsValue::Bool(isInterrupted));
+
+    EventsApi::SendEvent("sleepStop", { JsValue::Undefined(), obj });
+  });
+
+  return RE::BSEventNotifyControl::kContinue;
+}
+
+RE::BSEventNotifyControl GameEventSinks::ProcessEvent(
+  const RE::TESActorLocationChangeEvent* event,
+  RE::BSTEventSource<RE::TESActorLocationChangeEvent>* eventSource)
+{
+  auto converted =
+    reinterpret_cast<const TESEvents::TESActorLocationChangeEvent*>(event);
+
+  if (!converted) {
+    return RE::BSEventNotifyControl::kContinue;
+  }
+
+  auto actor = converted->actor.get();
+  auto oldLoc = converted->oldLoc;
+  auto newLoc = converted->newLoc;
+  auto actorId = actor ? actor->formID : 0;
+  auto oldLocId = oldLoc ? oldLoc->formID : 0;
+  auto newLocId = newLoc ? newLoc->formID : 0;
+
+  SkyrimPlatform::GetSingleton().AddUpdateTask(
+    [actor, actorId, oldLoc, oldLocId, newLoc, newLocId] {
+      auto obj = JsValue::Object();
+
+      auto actorLocal = RE::TESForm::LookupByID(actorId);
+      auto oldLocLocal = RE::TESForm::LookupByID(oldLocId);
+      auto newLocLocal = RE::TESForm::LookupByID(newLocId);
+
+      if (actorLocal == nullptr || oldLocLocal == nullptr ||
+          newLocLocal == nullptr ||
+          actorLocal->formType != RE::FormType::ActorCharacter ||
```

Comment:
No need to check form types here since we already check `actorLocal != actor || oldLocLocal != oldLoc ||
          newLocLocal != newLoc`